### DISCUSSION TITLE
Web app instantiation

### DISCUSSION THREAD

**Discussion by idlemind** [microsoft/semantic-kernel#7090] [Q&A] [answered]
I'm creating a web app chat client using c# asp.net MVC. What's the most efficient way of instantiating and instance of the kernel per chat session? I doubt instantiating a new instance per web request is a great idea? Also, each chat session will a differing set of plugins and kernel.Data. Thanks

**Comment by dmytrostruk** [comment, marked as answer]
Hi @idlemind !

> What's the most efficient way of instantiating and instance of the kernel per chat session? I doubt instantiating a new instance per web request is a great idea?

Kernel is a lightweight object, so it should be safe to initialize a new instance as Scoped (per request) or Transient (per object).

There is an extension method to easily add Kernel to your application, and we add it as Transient by default:
https://github.com/microsoft/semantic-kernel/blob/e77ffdf84a29311ac66ecb227c4421f10fb1fbad/dotnet/src/SemanticKernel.Abstractions/Services/KernelServiceCollectionExtensions.cs#L18-L29

In your application you can add it to DI container like this:
```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddKernel();
builder.Services.AddOpenAIChatCompletion(openAIOptions.ChatModelId, openAIOptions.ApiKey);
```

> Also, each chat session will a differing set of plugins and kernel.Data

If you will initialize Kernel instance per request or per object, it will have 0 imported plugins, so in each place you can import different set of plugins and have different `kernel.Data` property bag.

**Reply by idlemind** [reply]
Brilliant. Thanks for your reply!
